File: ПМИВС/lab5/Lab_5_Task_3_Shibitov_Nikolay/app/src/main/java/com/example/lab_5_task_3_shibitov_nikolay/MainActivity.java
package com.example.lab_5_task_3_shibitov_nikolay;

import android.app.AlertDialog;
import android.gesture.Gesture;
import android.gesture.GestureLibraries;
import android.gesture.GestureLibrary;
import android.gesture.GestureOverlayView;
import android.gesture.Prediction;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.Random;

public class MainActivity extends AppCompatActivity
        implements GestureOverlayView.OnGesturePerformedListener {

    private GestureLibrary gestureLibrary;
    private TextView tvPrompt;
    private TextView tvInput;
    private TextView tvResult;
    private StringBuilder inputBuffer = new StringBuilder();
    private int secretNumber;
    private boolean gameActive = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvPrompt = findViewById(R.id.tvPrompt);
        tvInput  = findViewById(R.id.tvInput);
        tvResult = findViewById(R.id.tvResult);

        // Load gesture library from raw resources
        gestureLibrary = GestureLibraries.fromRawResource(this, R.raw.gestures);
        if (!gestureLibrary.load()) {
            Toast.makeText(this, "Не удалось загрузить жесты", Toast.LENGTH_SHORT).show();
            Log.e("Gestures", "Error loading gestures from R.raw.gestures");
        }

        GestureOverlayView gestureOverlayView = findViewById(R.id.gestureOverlay);
        gestureOverlayView.addOnGesturePerformedListener(this);

        Button btnStart = findViewById(R.id.btnStart);
        btnStart.setOnClickListener(v -> startNewGame());

        Button btnClear = findViewById(R.id.btnClear);
        btnClear.setOnClickListener(v -> {
            inputBuffer.setLength(0);
            tvInput.setText("Число: ");
        });

        Button btnCheck = findViewById(R.id.btnCheck);
        btnCheck.setOnClickListener(v -> checkGuess());
    }

    private void startNewGame() {
        secretNumber = new Random().nextInt(100) + 1;
        gameActive = true;
        inputBuffer.setLength(0);
        tvPrompt.setText("Угадайте число от 1 до 100");
        tvInput.setText("Число: ");
        tvResult.setText("");
        Toast.makeText(this, "Игра началась!", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onGesturePerformed(GestureOverlayView overlay, Gesture gesture) {
        if (!gameActive) {
            Toast.makeText(this, "Нажмите «Старт»", Toast.LENGTH_SHORT).show();
            return;
        }

        ArrayList<Prediction> predictions = gestureLibrary.recognize(gesture);
        // Снижаем порог до 1.0 для лучшего распознавания
        if (!predictions.isEmpty() && predictions.get(0).score >= 1.0) {
            String name = predictions.get(0).name;
            if (name.matches("[0-9]")) {
                inputBuffer.append(name);
                tvInput.setText("Число: " + inputBuffer);
            }
        } else {
            Toast.makeText(this, "Жест не понятен", Toast.LENGTH_SHORT).show();
        }
    }

    private void checkGuess() {
        if (!gameActive) {
            Toast.makeText(this, "Начните новую игру", Toast.LENGTH_SHORT).show();
            return;
        }
        if (inputBuffer.length() == 0) {
            Toast.makeText(this, "Сначала введите число жестами", Toast.LENGTH_SHORT).show();
            return;
        }
        try {
            int guess = Integer.parseInt(inputBuffer.toString());
            inputBuffer.setLength(0);
            tvInput.setText("Число: ");

            if (guess < secretNumber) {
                tvResult.setText(guess + " — слишком мало");
            } else if (guess > secretNumber) {
                tvResult.setText(guess + " — слишком много");
            } else {
                gameActive = false;
                new AlertDialog.Builder(this)
                        .setTitle("Победа!")
                        .setMessage("Вы угадали число " + secretNumber + "!")
                        .setPositiveButton("Играть еще", (d, w) -> startNewGame())
                        .show();
            }
        } catch (NumberFormatException e) {
            Toast.makeText(this, "Ошибка формата", Toast.LENGTH_SHORT).show();
            inputBuffer.setLength(0);
        }
    }
}

File: ПМИВС/lab5/Lab_5_Task_4_Shibitov_Nikolay/app/src/main/java/com/example/lab_5_task_4_shibitov_nikolay/MainActivity.kt
package com.example.lab_5_task_4_shibitov_nikolay

import android.content.res.Configuration
import android.gesture.Gesture
import android.gesture.GestureLibraries
import android.gesture.GestureLibrary
import android.gesture.GestureOverlayView
import android.gesture.Prediction
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

class MainActivity : AppCompatActivity(),
        GestureOverlayView.OnGesturePerformedListener {

    private lateinit var tvDisplay: TextView
    private lateinit var tvResult: TextView
    private lateinit var tvGestureHint: TextView
    private lateinit var spinnerFrom: Spinner
    private lateinit var spinnerTo: Spinner
    private lateinit var btnGestureMode: Button

    private var inputString: String = "0"
    private var gestureMode: Boolean = false
    private lateinit var gestureLibrary: GestureLibrary

    // Conversion factors from Lab 2
    private val fromFactors = mapOf(
        "m²"  to 1.0,
        "cm²" to 0.0001,
        "km²" to 1_000_000.0,
        "ha"  to 10_000.0
    )
    private val toFactors = mapOf(
        "in²" to 0.00064516,
        "ft²" to 0.09290304,
        "yd²" to 0.83612736,
        "acre" to 4046.85642,
        "mi²" to 2_589_988.11
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvDisplay      = findViewById(R.id.tvDisplay)
        tvResult       = findViewById(R.id.tvResult)
        tvGestureHint  = findViewById(R.id.tvGestureHint)
        spinnerFrom    = findViewById(R.id.spinnerFrom)
        spinnerTo      = findViewById(R.id.spinnerTo)
        btnGestureMode = findViewById(R.id.btnGestureMode)

        if (savedInstanceState != null) {
            inputString = savedInstanceState.getString("inputString", "0")
            tvResult.text = savedInstanceState.getString("resultString", "")
        }
        updateDisplay()

        // Load gesture library
        gestureLibrary = GestureLibraries.fromRawResource(this, R.raw.gestures)
        if (!gestureLibrary.load()) {
            Toast.makeText(this, "Gesture library not found (create via Gesture Builder)",
                    Toast.LENGTH_LONG).show()
        }

        // Attach gesture listener
        val overlay = findViewById<GestureOverlayView>(R.id.gestureOverlay)
        overlay.addOnGesturePerformedListener(this)

        // Gesture mode toggle
        btnGestureMode.setOnClickListener {
            gestureMode = !gestureMode
            overlay.isEnabled = gestureMode
            if (gestureMode) {
                btnGestureMode.text = getString(R.string.gesture_mode_on)
                tvGestureHint.visibility = View.VISIBLE
            } else {
                btnGestureMode.text = getString(R.string.gesture_mode_off)
                tvGestureHint.visibility = View.GONE
            }
        }
        // Start with gesture overlay disabled (buttons are default input)
        overlay.isEnabled = false
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("inputString", inputString)
        outState.putString("resultString", tvResult.text.toString())
    }

    // --- Gesture input ---

    override fun onGesturePerformed(overlay: GestureOverlayView, gesture: Gesture) {
        val predictions: ArrayList<Prediction> = gestureLibrary.recognize(gesture)
        if (predictions.isEmpty() || predictions[0].score < 1.0) {
            Toast.makeText(this, "Gesture not recognized", Toast.LENGTH_SHORT).show()
            return
        }
        applyInput(predictions[0].name)
    }

    private fun applyInput(name: String) {
        when (name) {
            "0","1","2","3","4","5","6","7","8","9" -> {
                inputString = if (inputString == "0") name else inputString + name
            }
            "dot" -> {
                if (!inputString.contains(".")) inputString += "."
            }
            "clear" -> {
                inputString = "0"
                tvResult.text = ""
            }
            "backspace" -> {
                inputString = if (inputString.length > 1) {
                    val s = inputString.dropLast(1)
                    if (s == "-") "0" else s
                } else "0"
            }
            "plus_minus" -> {
                inputString = if (inputString.startsWith("-")) {
                    inputString.drop(1)
                } else {
                    if (inputString == "0") "0" else "-$inputString"
                }
            }
            "equals" -> calculateResult()
            else -> Toast.makeText(this, "Unknown gesture: $name", Toast.LENGTH_SHORT).show()
        }
        updateDisplay()
    }

    // --- Button input (original Lab 2 logic, unchanged) ---

    @Suppress("UNUSED_PARAMETER")
    fun onButtonClick(view: View) {
        val text = (view as Button).text.toString()
        when (text) {
            "C"  -> { inputString = "0"; tvResult.text = "" }
            "⌫"  -> {
                inputString = if (inputString.length > 1) {
                    val s = inputString.dropLast(1)
                    if (s == "-") "0" else s
                } else "0"
            }
            "±"  -> {
                inputString = if (inputString.startsWith("-")) {
                    inputString.drop(1)
                } else {
                    if (inputString == "0") "0" else "-$inputString"
                }
            }
            "."  -> { if (!inputString.contains(".")) inputString += "." }
            "="  -> calculateResult()
            else -> { // digits
                inputString = if (inputString == "0") text else inputString + text
            }
        }
        updateDisplay()
    }

    @Suppress("DEPRECATION", "UNUSED_PARAMETER")
    fun onLangClick(view: View) {
        val currentLang = resources.configuration.locales[0].language
        val newLang = if (currentLang == "ru") "en" else "ru"
        val locale = Locale(newLang)
        Locale.setDefault(locale)
        val config = Configuration(resources.configuration)
        config.setLocale(locale)
        resources.updateConfiguration(config, resources.displayMetrics)
        recreate()
    }

    // --- Shared logic ---

    private fun updateDisplay() {
        tvDisplay.text = inputString
    }

    private fun calculateResult() {
        val value = inputString.toDoubleOrNull()
        if (value == null) {
            tvResult.text = getString(R.string.error_invalid)
            return
        }
        val fromUnit = spinnerFrom.selectedItem.toString()
        val toUnit   = spinnerTo.selectedItem.toString()
        val factorFrom = fromFactors[fromUnit] ?: 1.0
        val factorTo   = toFactors[toUnit]   ?: 1.0
        if (factorTo == 0.0) {
            tvResult.text = getString(R.string.error_div_zero)
            return
        }
        val result = (value * factorFrom) / factorTo
        val df = DecimalFormat("0.######", DecimalFormatSymbols(Locale.US))
        tvResult.text = df.format(result)
    }
}
